### ISSUE TITLE
Follow web-browser authentication flow on different machine

### ISSUE THREAD

**Issue by ckhawks** [streambinder/spotitube#42]
I'm running Ubuntu 18.04.2 headless on a VM, this is currently Go-lang 1.7.6, but I tried the latest 1.12.5 as well. ffmpeg and youtube-dl are installed, it seems like it's failing trying to open the web server for authentication. Not sure what to do, using the prebuilt .deb. Thanks!

```
stlrc@spotitubevm:~/spotitube$ spotitube -playlist spotify:user:meltedexistence:playlist:6J3Jhg7vKC3PHurFUyigx3 -disable-gui
Folder: ~/spotitube
Version: 24
Caller: installed
Date: 2019-05-14 03:35:56
URL: https://github.com/streambinder/spotitube
License: GPLv2

Version youtube-dl: 2019.05.11

Version ffmpeg: 3.4.6
Authentication URL: http://tinyurl.com/y4mctlf2
Unable to authenticate to spotify.
Authentication completed.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x82f6893]

goroutine 1 [running]:
github.com/zmb3/spotify.(*Client).CurrentUser(0x0, 0x19, 0x83c3ec0, 0xa2a6070)
        /home/streambinder/Workspace/spotitube/src/github.com/zmb3/spotify/user.go:87 +0x33
spotify.(*Spotify).User(0xa06f4d0, 0x844711c, 0x19, 0x80, 0x0)
        /home/streambinder/Workspace/spotitube/src/spotify/interface.go:78 +0x23
main.mainFetch()
        /home/streambinder/Workspace/spotitube/src/main/main.go:218 +0x18d
main.main()
        /home/streambinder/Workspace/spotitube/src/main/main.go:204 +0xf5b
```

**Comment by streambinder** [comment]
I can fix the panic error pretty easily, but here the issue is another one: how have you opened that URL (`http://tinyurl.com/y4mctlf2`)?

**Comment by ckhawks** [comment]
I was opening it on my personal computer in Windows, I set the callback url to http://theinternalipofthevm:8080/callback, but it would not connect because (I believe?) the web server was dying before I could send the request to it

**Comment by streambinder** [comment]
The problem is about the fact `callback_url` doesn't get altered if not modified directly on the Spotify app dashboard (and in SpotiTube case, the only authorized `callback_url`s are `http://127.0.0.1:8080/callback` and `http://localhost:8080/callback`, which should be the same).
SpotiTube opens up a socket on `localhost:8080` and Spotify, when authentication succeeds, redirects on `callback_url` (which by default is `http://localhost:8080/callback`).
The issue you're reporting denounces a new use-scenario: run the application on a machine which isn't able to directly handle the user authentication flow and let the user handle it using a web browser on another machine.
Basically the steps to enable this could be:
1. enable a dedicated generic `callback_url` as whitelisted Spotify application Callback URL (e.g. `http://spotitube.local:8080/callback`)
2. enable a new application flag to tell spoptitube to:
   1. open up the socket on all network interfaces: `0.0.0.0:8080`
   2. set `http://spotitube.local:8080/callback` as `callback_url`
   3. prompt the user to add `spotitube.local` to his `hosts` file (on the machine which is handling the authentication flow - your Windows 10 machine), pointing to the effective machine IP which spotitube is running on (your Ubuntu headless machine).

**Comment by streambinder** [comment]
@ckhawks, please, retry with this binary (it's a binary, `.txt` extension is used to let GitHub upload it on this issue - remove it), appending `-authenticate-outside` flag: [spotitube](https://github.com/streambinder/spotitube/files/3183049/spotitube.txt)

**Comment by ckhawks** [comment]
Took me a minute but I think it's working. Thank you so much man, I've been trying to download from Spotify for ages. From my understanding, if I run it without any -playlist or specification, it is just trying to download my saved songs. How does this handle large libraries (+5k songs)? <3

**Comment by streambinder** [comment]
> Took me a minute but I think it's working. Thank you so much man, I've been trying to download from Spotify for ages. From my understanding, if I run it without any -playlist or specification, it is just trying to download my saved songs. How does this handle large libraries (+5k songs)? <3

You're right, without `-playlist` flag it automatically switch to the library download.
My library is ~1.5k and gets correctly synchronized. Try yourself and let me know if it can handle also ~5k ones.

> oh also, I'm assuming we have to re-authenticate every time? Not a huge problem, because I assume it's technically a lot harder to make it save, but just wondering. Thanks!

I'm working on a patch to make SpotiTube wrap and use webdrivers to batch authenticate. If this will be ever working, the whole authentication-via-browser phase will be dropped (or, at least, used as fallback method). In the meantime, yup, you'll need to authenticate everytime.

> Ran into another issue. If the playlist contains local files the script just dies.

Wow. Never encountered such use case.
Would you please open another issue and attach the log provided by spotitube with `-log` flag when executed to retrigger the issue?
Thank you.

**Comment by ckhawks** [comment]
I tried to download both a playlist with only a local file and a playlist with local file + normal Spotify song and it states
```
┌─ SPOTITUBE ─────────────────────────────────┐┌─ STATUS ──────────────────────────────────────────────────────────────────────────────────┐
│ Folder: /downloads                          ││ Indexing started...                                                                       │
│ Log: spotitube_2019-05-20_08.20.49.log      ││ Indexing finished: 3 tracks indexed and 3 linked.                                         │
│ Version youtube-dl: 2019.05.11              ││ Authentication URL: http://tinyurl.com/y3ns3j6n                                           │
│ Version ffmpeg: 3.4.6                       ││ Authentication completed.                                                                 │
│ Session user: meltedexistence               ││ Fetching playlist data...                                                                 │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
└──────────────────────────────────────────┌───────────────────────────────────────────────────┐                                           │
┌─ INFORMATIONS ───────────────────────────│Something went wrong while fetching playlist info. │                                           │
│ Version: 24                              └───────────────────────────────────────────────────┘                                           │
│ Date: 2019-05-20 08:20:49                   ││                                                                                           │
│ Caller: installed                           ││                                                                                           │
│ URL: https://github.com/streambinder/spotitu││                                                                                           │
│ License: GPLv2                              ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
│                                             ││                                                                                           │
└─────────────────────────────────────────────┘└───────────────────────────────────────────────────────────────────────────────────────────┘
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                                                          │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

**Comment by streambinder** [comment]
Thank you.
Let me know if synchronizing huge libraries/playlist (several thousands of tracks) is working as flawlessly as always.
Flag `-authenticate-outside` will be soon merged into master branch, anyway. Once done, I will close this issue.

**Comment by streambinder** [comment]
Please, test this binary (it applied the `-authenticate-outside` patch over the new master branch additions): 
[spotitube.txt](https://github.com/streambinder/spotitube/files/3199477/spotitube.txt).
Let me know.
